### PR TITLE
MF-1 - Add configuration file

### LINKED ISSUE #1: Add configuration file
Things and Channels created by provision requests should be configurable using a config file. We can specify the number of Things and Channels that are going to be created, its names, and relations.

### PR DESCRIPTION
**Pull request by mteodor**
Resolves #1

### CODE REVIEW

File: cmd/main.go [hunk 1]
```diff
@@ -7,20 +7,21 @@ import (
 	"os"
 	"os/signal"
 	"strconv"
-	"strings"
 	"syscall"
 
 	"github.com/mainflux/mainflux"
 	"github.com/mainflux/mainflux/logger"
-	mfSDK "github.com/mainflux/mainflux/sdk/go"
+	mfsdk "github.com/mainflux/mainflux/sdk/go"
```
Comment (drasko) [inline cmd/main.go:14]: Keep SDK capitalized

File: cmd/main.go [hunk 2]
```diff
@@ -60,28 +64,45 @@ const (
...
 	}
+	logger, err := logger.New(os.Stdout, cfg.Server.LogLevel)
+	if err != nil {
+		log.Fatalf(err.Error())
+	}
+	cfg, err = loadConfigFromFile(cfg, logger)
+	if err != nil {
+		logger.Warn("Continue with settings from env vars")
+	}
+	if err == nil {
```
Comment (drasko) [inline cmd/main.go:84]: Why is this check needed?

Comment (mteodor) [reply cmd/main.go:88]: there is a message that is logged and it should explain why

first we create config from environment variables
and than try to load config from file 
if load from config file fails we use config created from envs
if not we use config from file, we log both

Comment (drasko) [reply cmd/main.go:88]: Please re-check these `if`s, second one is redundant.

File: api/responses.go [hunk 3]
```diff
@@ -7,12 +7,12 @@ import (
 )
 
 type provisionRes struct {
-	Thing       sdk.Thing     `json:"thing"`
-	Channels    []sdk.Channel `json:"channels"`
-	ClientCert  string        `json:"client_cert,omitempty"`
-	ClientKey   string        `json:"client_key,omitempty"`
-	CACert      string        `json:"ca_cert,omitempty"`
-	Whitelisted bool          `json:"whitelisted,omitempty"`
+	Things      []sdk.Thing       `json:"thing"`
```
Comment (manuio) [inline api/responses.go:10]: Fix ``json:"things"``

**Comment by nmarcetic** [comment]
LGTM!

File: cmd/main.go [hunk 4]
```diff
@@ -98,74 +120,101 @@ func main() {
...
-	}
-	thingsSDK := mfSDK.NewSDK(thingSdkCfg)
 
-	userSdkCfg := mfSDK.Config{
-		BaseURL:           mainflux.Env(envUsersLocation, defUsersLocation),
-		HTTPAdapterPrefix: "http",
-		MsgContentType:    "application/json",
-		TLSVerification:   tls,
+		// This is default conf for provision if there is no
+		// config file
```
Comment (drasko) [inline cmd/main.go:193]: Make this comment one line long, no need to break in 2 lines (it is more readable)

File: cmd/main.go [hunk 5]
```diff
@@ -60,28 +64,46 @@ const (
...
+
+	thingSdkCfg := mfSDK.Config{
+		BaseURL:           mainflux.Env(envThingsLocation, defThingsLocation),
+		HTTPAdapterPrefix: "http",
+		MsgContentType:    "application/json",
+		TLSVerification:   cfg.Server.TLS,
+	}
+	thingsSDK := mfSDK.NewSDK(thingSdkCfg)
+
+	userSdkCfg := mfSDK.Config{
```
Comment (drasko) [inline cmd/main.go:97]: SDK

File: cmd/main.go [hunk 6]
```diff
@@ -60,28 +64,46 @@ const (
...
+	logger, err := logger.New(os.Stdout, cfg.Server.LogLevel)
 	if err != nil {
 		log.Fatalf(err.Error())
 	}
+	cfg, err = loadConfigFromFile(cfg, logger)
+	if err != nil {
+		logger.Warn("Continue with settings from env vars")
+	}
+
+	thingSdkCfg := mfSDK.Config{
```
Comment (drasko) [inline cmd/main.go:89]: SDK

File: cmd/main.go [hunk 7]
```diff
@@ -60,28 +64,46 @@ const (
...
+	thingsSDK := mfSDK.NewSDK(thingSdkCfg)
+
+	userSdkCfg := mfSDK.Config{
+		BaseURL:           mainflux.Env(envUsersLocation, defUsersLocation),
+		HTTPAdapterPrefix: "http",
+		MsgContentType:    "application/json",
+		TLSVerification:   cfg.Server.TLS,
+	}
+	userSDK := mfSDK.NewSDK(userSdkCfg)
+	sdk := provsdk.New(cfg.Server.MfCertsURL, cfg.Server.MfBSURL, cfg.Server.MfWhiteListURL, thingsSDK, userSDK)
```
Comment (drasko) [inline cmd/main.go:104]: SDK

File: sdk/sdk_test.go [hunk 8]
```diff
@@ -66,7 +66,7 @@ func delete(rw http.ResponseWriter, r *http.Request) {
 }
 
 func createToken(rw http.ResponseWriter, r *http.Request) {
-	u := mfsdk.User{}
+	u := mfSDK.User{}
```
Comment (manuio) [inline sdk/sdk_test.go:69]: You can use `var u mfSDK.User` here

File: config.go [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+	MfBSURL        string `toml:"mf_bs_url"`
+	MfWhiteListURL string `toml:"mf_whit_list"`
+	MfCertsURL     string `toml:"mf_certs_url"`
+}
+
+type Bootstrap struct {
+	X509Provision bool   `toml:"provision_certs"`
+	BSProvision   bool   `toml:"provision_bs"`
+	AutoWhiteList bool   `toml:"bs_autowhite_list"`
+	BSContent     string `toml:"bs_content"`
```
Comment (manuio) [inline config.go:38]: This fields and toml tags are not consistent

File: sdk/sdk_test.go [hunk 10]
```diff
@@ -84,7 +84,7 @@ func createToken(rw http.ResponseWriter, r *http.Request) {
 }
 
 func createThing(rw http.ResponseWriter, r *http.Request) {
-	t := mfsdk.Thing{}
+	t := mfSDK.Thing{}
```
Comment (manuio) [inline sdk/sdk_test.go:87]: Use `var` here

File: sdk/sdk_test.go [hunk 11]
```diff
@@ -113,13 +113,13 @@ func thing(rw http.ResponseWriter, r *http.Request) {
...
 	}
-	data, _ := json.Marshal(mfsdk.Thing{})
+	data, _ := json.Marshal(mfSDK.Thing{})
 	rw.WriteHeader(http.StatusOK)
 	rw.Write(data)
 }
 
 func createChannel(rw http.ResponseWriter, r *http.Request) {
-	c := mfsdk.Channel{}
+	c := mfSDK.Channel{}
```
Comment (manuio) [inline sdk/sdk_test.go:122]: Use `var` here

File: sdk/sdk_test.go [hunk 12]
```diff
@@ -180,7 +180,7 @@ func saveToBootstrap(rw http.ResponseWriter, r *http.Request) {
 	if !auth(rw, r) || !ct(rw, r) {
 		return
 	}
-	cfg := provsdk.BSConfig{}
+	cfg := provSDK.BSConfig{}
```
Comment (manuio) [inline sdk/sdk_test.go:183]: Use `var` here

File: service.go [hunk 13]
```diff
@@ -16,157 +28,140 @@ type Service interface {
...
+		Channels: channels,
 	}
 
-	if ps.x509Provision {
-		certs, err := ps.sdk.Cert(thingCreated.ID, thingCreated.Key, token)
-		if err != nil {
-			return res, provsdk.ErrCerts
+	res.Whitelisted = make(map[string]bool, 0)
+	res.ClientCert = make(map[string]string, 0)
+	res.ClientKey = make(map[string]string, 0)
```
Comment (manuio) [inline service.go:123]: If Map is empty you can do:
```
	var res.Whitelisted = map[string]bool{}
	var res.ClientCert = map[string]string{}
	var res.ClientKey = map[string]string{}
```

File: service.go [hunk 14]
```diff
@@ -16,157 +28,140 @@ type Service interface {
...
+	certs := provSDK.Cert{}
```
Comment (manuio) [inline service.go:124]: Use `var` here

File: service.go (hunk 13 above)
Comment (mteodor) [reply service.go:123]: why ?

Comment (mteodor) [reply service.go:123]: wait, I cant do that, this are struct members

Comment (drasko) [reply service.go:123]: Why are these `Res` struct members not initialized on line https://github.com/mainflux/provision/pull/5/files#diff-30b92219c8fd27dd239b4dcd4bb4d838R116 during the `res` declaration?

Comment (manuio) [reply service.go:123]: right, my bad. The var fashion :)

File: config.go [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+	MfBSURL        string `toml:"mf_bs_url"`
+	MfWhiteListURL string `toml:"mf_whit_list"`
+	MfCertsURL     string `toml:"mf_certs_url"`
+}
+
+type Bootstrap struct {
+	X509Provision bool   `toml:"provision_certs"`
+	BSProvision   bool   `toml:"provision"`
+	AutoWhiteList bool   `toml:"autowhite_list"`
+	BSContent     string `toml:"content"`
```
Comment (manuio) [inline config.go:38]: Why we don't follow the structure naming?

Comment (drasko) [reply config.go:38]: Should be aligned.

Comment (mteodor) [reply config.go]: should be resolved

**Pull request merged**